File: backend/app/services/tool_analyzer.py
from __future__ import annotations

from typing import Any

from app.schemas.phase4 import ToolPermissionReport

BUILT_IN_TOOLS = {
    "bash",
    "edit",
    "write",
    "read",
    "glob",
    "grep",
    "webfetch",
    "websearch",
    "skill",
    "todowrite",
    "question",
    "lsp",
    "apply_patch",
    "task",
}

SECURITY_SENSITIVE_TOOLS = {"bash", "edit", "write", "apply_patch", "webfetch"}

PERMISSION_VALUES = {"allow", "ask", "deny"}


def analyze_tools(permissions: dict[str, Any] | None) -> list[ToolPermissionReport]:
    """Analyze tool permission entries from an opencode.json config."""
    if not permissions or not isinstance(permissions, dict):
        return []

    tools: list[ToolPermissionReport] = []

    for key, value in permissions.items():
        # Skip "bash" if value is a dict (sub-permission rules)
        if isinstance(value, dict):
            tools.append(
                ToolPermissionReport(
                    tool_name=key,
                    permission=None,
                    is_wildcard="*" in key,
                    is_security_sensitive=key in SECURITY_SENSITIVE_TOOLS or key == "bash",
                    recommendation="Sub-permission rules detected — review individual entries",
                )
            )
            continue

        perm = str(value) if isinstance(value, str) else None
        is_wildcard = "*" in key
        is_sensitive = (
            key in SECURITY_SENSITIVE_TOOLS or key in BUILT_IN_TOOLS and key.startswith("bash")
        )

        recommendation: str | None = None
        if perm and perm not in PERMISSION_VALUES:
            recommendation = f"'{perm}' is not a valid permission (use allow/ask/deny)"
        elif is_sensitive and perm == "allow":
            recommendation = f"'{key}' is security-sensitive — consider changing to 'ask'"
        elif is_sensitive and perm is None:
            recommendation = f"'{key}' is security-sensitive — add an explicit permission"

        tools.append(
            ToolPermissionReport(
                tool_name=key,
                permission=perm if perm in PERMISSION_VALUES else perm,
                is_wildcard=is_wildcard,
                is_security_sensitive=is_sensitive,
                recommendation=recommendation,
            )
        )

    return tools


def get_missing_critical(tools: list[ToolPermissionReport]) -> list[str]:
    """Identify security-sensitive tools that have no explicit permission rule."""
    covered = {t.tool_name for t in tools}
    missing: list[str] = []
    for tool in sorted(SECURITY_SENSITIVE_TOOLS):
        # Check if tool is covered by any rule (exact or wildcard)
        covered_by_wildcard = any(
            name.endswith("*") and tool.startswith(name[:-1]) for name in covered
        )
        if tool not in covered and not covered_by_wildcard:
            missing.append(tool)
    return missing

File: frontend/src/lib/services/tool-analyzer.ts
import type { ToolPermissionReport } from "@/types/phase4"

const BUILT_IN_TOOLS = new Set(["bash", "edit", "write", "read", "glob", "grep", "webfetch", "websearch", "skill", "todowrite", "question", "lsp", "apply_patch", "task"])
const SECURITY_SENSITIVE_TOOLS = new Set(["bash", "edit", "write", "apply_patch", "webfetch"])
const PERMISSION_VALUES = new Set(["allow", "ask", "deny"])

export function analyzeTools(permissions: Record<string, unknown> | null): {
  tools: ToolPermissionReport[]
  missingCritical: string[]
  overallScore: number
} {
  if (!permissions || typeof permissions !== "object") {
    return { tools: [], missingCritical: [], overallScore: 0 }
  }

  const tools: ToolPermissionReport[] = []

  for (const [key, value] of Object.entries(permissions)) {
    if (typeof value === "object" && value !== null && !Array.isArray(value)) {
      tools.push({
        tool_name: key,
        permission: null,
        is_wildcard: key.includes("*"),
        is_security_sensitive: key === "bash" || SECURITY_SENSITIVE_TOOLS.has(key),
        recommendation: "Sub-permission rules detected — review individual entries",
      })
      continue
    }

    const perm = typeof value === "string" ? value : null
    const isWildcard = key.includes("*")
    const isSensitive = SECURITY_SENSITIVE_TOOLS.has(key) || (BUILT_IN_TOOLS.has(key) && key.startsWith("bash"))

    let recommendation: string | null = null
    if (perm && !PERMISSION_VALUES.has(perm)) {
      recommendation = `'${perm}' is not a valid permission (use allow/ask/deny)`
    } else if (isSensitive && perm === "allow") {
      recommendation = `'${key}' is security-sensitive — consider changing to 'ask'`
    } else if (isSensitive && perm === null) {
      recommendation = `'${key}' is security-sensitive — add an explicit permission`
    }

    tools.push({
      tool_name: key,
      permission: perm && PERMISSION_VALUES.has(perm) ? perm : perm,
      is_wildcard: isWildcard,
      is_security_sensitive: isSensitive,
      recommendation,
    })
  }

  const covered = new Set(tools.map((t) => t.tool_name))
  const missingCritical: string[] = []
  for (const tool of [...SECURITY_SENSITIVE_TOOLS].sort()) {
    const coveredByWildcard = [...covered].some((name) => name.endsWith("*") && tool.startsWith(name.slice(0, -1)))
    if (!covered.has(tool) && !coveredByWildcard) missingCritical.push(tool)
  }

  const overallScore = tools.length > 0
    ? Math.max(0, 100 - missingCritical.length * 20 - tools.filter((t) => t.recommendation && t.recommendation.includes("not a valid permission")).length * 15)
    : 0

  return { tools, missingCritical, overallScore }
}
